include/os/linux/spl/sys/cred.h

```diff
@@ -45,32 +45,73 @@ typedef struct cred cred_t;
 #define	SGID_TO_KGID(x)		(KGIDT_INIT(x))
 #define	KGIDP_TO_SGIDP(x)	(&(x)->val)
 
-static inline uid_t zfs_uid_into_mnt(struct user_namespace *mnt_ns, uid_t uid)
+extern boolean_t zfs_is_init_userns(struct user_namespace *ns);
+extern struct user_namespace *zfs_get_init_userns(void);
+
+static inline struct user_namespace *zfs_i_user_ns(struct inode *inode)
+{
+#ifdef HAVE_SUPER_USER_NS
+	return (inode->i_sb->s_user_ns);
+#else
+	return (NULL);
+#endif
+}
+
+static inline boolean_t zfs_no_idmapping(struct user_namespace *mnt_userns,
+    struct user_namespace *fs_userns)
 {
-	if (mnt_ns)
-		return (__kuid_val(make_kuid(mnt_ns, uid)));
-	return (uid);
+	return (!mnt_userns || !fs_userns ||
```

Comment:
I kind of agree but here is my take on having NULL - on older kernel version(< 5.12) which does not provide "mnt_userns" in the inode_operations callbacks, a NULL mnt_userns would provide a shortcut for no id mapping. If we try to use &init_user_ns in such case, a function call is needed (due to its GPL nature) and we have to check if it is the initial user ns in order to have a shortcut. As you can see, this would need more CPU cycles.

@behlendorf What is your thought on this?